File: AllJoyn/Samples/ZigBeeAdapter/AdapterLib/LSFHandler.cs
// Copyright (c) 2015, Microsoft Corporation
//
// Permission to use, copy, modify, and/or distribute this software for any
// purpose with or without fee is hereby granted, provided that the above
// copyright notice and this permission notice appear in all copies.
//
// THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
// WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
// MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY
// SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
// WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
// ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF OR
// IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
//

using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using BridgeRT;

namespace AdapterLib
{
    class LSFHandler : ILSFHandler
    {
        private const int OEM_LS_COLOR_TEMPERATURE_MIN = 2700;
        private const int OEM_LS_COLOR_TEMPERATURE_MAX = 9000;

        private const uint ERROR_SUCCESS = 0;

        // ILSFHandler
        //---------------
        public bool LampDetails_Color { get; }
        public uint LampDetails_ColorRenderingIndex { get; }
        public bool LampDetails_Dimmable { get; }
        public bool LampDetails_HasEffects { get; }
        public uint LampDetails_IncandescentEquivalent { get; }
        public uint LampDetails_LampBaseType { get; }
        public uint LampDetails_LampBeamAngle { get; }
        public string LampDetails_LampID { get; }
        public uint LampDetails_LampType { get; }
        public uint LampDetails_Make { get; }
        public uint LampDetails_MaxLumens { get; }
        public uint LampDetails_MaxTemperature { get; }
        public uint LampDetails_MaxVoltage { get; }
        public uint LampDetails_MinTemperature { get; }
        public uint LampDetails_MinVoltage { get; }
        public uint LampDetails_Model { get; }
        public uint LampDetails_Type { get; }
        public bool LampDetails_VariableColorTemp { get; }
        public uint LampDetails_Version { get; }
        public uint LampDetails_Wattage { get; }
        public uint LampParameters_BrightnessLumens { get; }
        public uint LampParameters_EnergyUsageMilliwatts { get; }
        public uint LampParameters_Version { get; }
        public uint[] LampService_LampFaults { get; } = { 0 };
        public uint LampService_LampServiceVersion { get; }
        public uint LampService_Version { get; }
        public uint LampState_ColorTemp { get; set; }
        public uint LampState_Hue { get; set; }
        public IAdapterSignal LampState_LampStateChanged { get; }
        public uint LampState_Saturation { get; set; }
        public uint LampState_Version { get; }

        // default on/off to off, will be updated later
        private bool m_onOffState = false;
        public bool LampState_OnOff
        {
            get
            {
                return GetOnOffStateFromDevice();
            }
            set
            {
                SetOnOffStateOnDevice(value);
            }
        }

        byte m_level = 0;
        public uint LampState_Brightness
        {
            get
            {
                return GetLevelFromDevice();
            }
            set
            {
                SetLevelOnDevice(value);
            }
        }
        public uint ClearLampFault(uint InLampFaultCode, out uint LampResponseCode, out uint OutLampFaultCode)
        {
            LampResponseCode = 0;
            OutLampFaultCode = InLampFaultCode;
            return ERROR_SUCCESS;
        }
        public uint LampState_ApplyPulseEffect(State FromState, State ToState, uint Period, uint Duration, uint NumPulses, ulong Timestamp, out uint LampResponseCode)
        {
            LampResponseCode = 0;
            return ERROR_SUCCESS;
        }
        public uint TransitionLampState(ulong Timestamp, State NewState, uint TransitionPeriod, out uint LampResponseCode)
        {
            LampResponseCode = 0;
            return ERROR_SUCCESS;
        }

        // ZigBee adapter related
        //------------------------

        private ZigBeeEndPoint m_endPoint = null;
        public LSFHandler(ZigBeeEndPoint endPoint, ZigBeeProfileLibrary.DeviceType type)
        {
            m_endPoint = endPoint;

            LampDetails_Version = 1;
            LampService_LampServiceVersion = 1;
            LampState_Version = 1;
            LampParameters_Version = 1;

            if(type == ZigBeeProfileLibrary.DeviceType.dimmableLight)
            {
                LampDetails_Dimmable = true;
            }
            else
            {
                LampDetails_Dimmable = false;
            }

            // only on/off and dimmable for now
            LampDetails_Color = false;
            LampDetails_VariableColorTemp = false;
            LampDetails_HasEffects = false;

            LampDetails_IncandescentEquivalent = 60;
            LampDetails_LampBaseType = (uint)BridgeRT.LSFLampBaseType.BASETYPE_E26;
            LampDetails_LampBeamAngle = 160;
            LampDetails_LampType = (uint)BridgeRT.LSFLampType.LAMPTYPE_A19;
            LampDetails_Make = (uint)BridgeRT.LSFLampMake.MAKE_OEM1;
            LampDetails_MaxLumens = 620;
            LampDetails_MaxTemperature = OEM_LS_COLOR_TEMPERATURE_MAX;
            LampDetails_MaxVoltage = 120;
            LampDetails_MinTemperature = OEM_LS_COLOR_TEMPERATURE_MIN;
            LampDetails_MinVoltage = 100;
            LampDetails_Model = 1;
            LampDetails_Type = (uint)BridgeRT.LSFDeviceType.TYPE_LAMP;
            LampDetails_VariableColorTemp = false;
            LampDetails_Wattage = 9;

            LampDetails_LampID = m_endPoint.SerialNumber;

            LampState_ColorTemp = 0;
            LampState_Hue = 0;
            LampState_Saturation = 0;

            // ZigBee light (on/off cluster) has no state change 
            LampState_LampStateChanged = null;
        }

        private bool GetOnOffStateFromDevice()
        {
            ZclAttribute attributeOnOff = null;
            System.Object value;

            // look for OnOff cluster of this end point and get the value of the on/off attribute
            var onOffCluster = m_endPoint.GetCluster(OnOffCluster.CLUSTER_ID);
            if (onOffCluster != null &&
                onOffCluster.InternalAttributeList.TryGetValue(OnOffCluster.ATTRIBUTE_ONOFF, out attributeOnOff) &&
                attributeOnOff.Read(out value) &&
                value is bool)
            {
                m_onOffState = (bool) value;
            }

            return m_onOffState;
        }

        private void SetOnOffStateOnDevice(bool newOnOffState)
        {
            ZclCommand command = null;

            // look for OnOff cluster of this end point
            var onOffCluster = m_endPoint.GetCluster(OnOffCluster.CLUSTER_ID);
            if (onOffCluster == null)
            {
                return;
            }

            // get on or off command depending on the on/off state to set
            if(newOnOffState)
            {
                onOffCluster.CommandList.TryGetValue(OnOffCluster.COMMAND_ON, out command);
            }
            else
            {
                onOffCluster.CommandList.TryGetValue(OnOffCluster.COMMAND_OFF, out command);
            }

            if (command == null)
            {
                return;
            }

            // send the command
            command.Send();
            m_onOffState = newOnOffState;
        }

        uint GetLevelFromDevice()
        {
            if(LampDetails_Dimmable)
            {
                ZclAttribute attributeLevel = null;
                System.Object value;

                // look for level control cluster of this end point and get the value of the current level attribute
                var levelControlCluster = m_endPoint.GetCluster(LevelControlCluster.CLUSTER_ID);
                if (levelControlCluster != null &&
                    levelControlCluster.InternalAttributeList.TryGetValue(LevelControlCluster.ATTRIBUTE_CURRENTLEVEL, out attributeLevel) &&
                    attributeLevel.Read(out value) &&
                    value is byte)
                {
                    m_level = (byte)value;
                }
            }

            return (uint)m_level;
        }

        void SetLevelOnDevice(uint newLevel)
        {
            ZclCommand command = null;

            // look for level control cluster and move to level command
            var levelControlCluster = m_endPoint.GetCluster(LevelControlCluster.CLUSTER_ID);
            if (levelControlCluster == null ||
                !levelControlCluster.CommandList.TryGetValue(LevelControlCluster.COMMAND_MOVETOLEVEL, out command))
            {
                return;
            }

            // get level parameter
            var parameter = command.GetInputParamByName(LevelControlCluster.LEVEL_PARAM);
            if(parameter == null)
            {
                return;
            }

            // level is a byte in ZCL and a uint32 in LSF 
            if(newLevel > LevelControlCluster.MAX_LEVEL)
            {
                newLevel = LevelControlCluster.MAX_LEVEL;
            }
            parameter.Data = (byte) newLevel;
            command.Send();
            m_level = (byte) newLevel;
        }
    }
}

File: AllJoyn/Samples/MockAdapter/AdapterLib/LSFHandler.cpp
#include "pch.h"
#include "LSFHandler.h"
#include "MockLampConsts.h"
#include "MockLampDevice.h"
#include "MockAdapter.h"

using namespace AdapterLib;
using namespace BridgeRT;

LSFHandler::LSFHandler(MockLampDevice^ parentDevice)
    : m_parentDevice(parentDevice)
{
}


LSFHandler::~LSFHandler()
{
}


// LampService Interface
uint32 
LSFHandler::LampService_Version::get()
{
    return 1;
}


uint32 
LSFHandler::LampService_LampServiceVersion::get()
{
    return 1;
}


Platform::Array<uint32>^
LSFHandler::LampService_LampFaults::get()
{
    Platform::Array<uint32>^ lampFaults = ref new Platform::Array<uint32>(1);
    lampFaults[0] = 0;
    return lampFaults;
}


uint32 
LSFHandler::ClearLampFault(
    _In_ uint32 InLampFaultCode,
    _Out_ uint32 *LampResponseCode,
    _Out_ uint32 *OutLampFaultCode
    )
{
    *LampResponseCode = 0;
    *OutLampFaultCode = InLampFaultCode;
    return ERROR_SUCCESS;
}

// LampParameters Interface
uint32 
LSFHandler::LampParameters_Version::get()
{
    return 1;
}


uint32 
LSFHandler::LampParameters_EnergyUsageMilliwatts::get()
{
    return 600;
}


uint32 
LSFHandler::LampParameters_BrightnessLumens::get()
{
    return 1000;
}


// LampDetails Interface
uint32 
LSFHandler::LampDetails_Version::get()
{
    return 1;
}


uint32 
LSFHandler::LampDetails_Make::get()
{
    return static_cast<uint32>(BridgeRT::LSFLampMake::MAKE_OEM1);
}


uint32 
LSFHandler::LampDetails_Model::get()
{
    return static_cast<uint32>(BridgeRT::LSFLampModel::MODEL_LED);
}


uint32 
LSFHandler::LampDetails_Type::get()
{
    return static_cast<uint32>(BridgeRT::LSFLampType::LAMPTYPE_A15);
}


uint32 
LSFHandler::LampDetails_LampType::get()
{
    return static_cast<uint32>(BridgeRT::LSFLampType::LAMPTYPE_A19);
}


uint32 
LSFHandler::LampDetails_LampBaseType::get()
{
    return static_cast<uint32>(BridgeRT::LSFLampBaseType::BASETYPE_E26);
}


uint32 
LSFHandler::LampDetails_LampBeamAngle::get()
{
    return 160;
}


bool 
LSFHandler::LampDetails_Dimmable::get()
{
    return false;
}


bool 
LSFHandler::LampDetails_Color::get()
{
    return false;
}


bool 
LSFHandler::LampDetails_VariableColorTemp::get()
{
    return false;
}


bool 
LSFHandler::LampDetails_HasEffects::get()
{
    return false;
}


uint32 
LSFHandler::LampDetails_MinVoltage::get()
{
    return 100;
}


uint32 
LSFHandler::LampDetails_MaxVoltage::get()
{
    return 120;
}


uint32 
LSFHandler::LampDetails_Wattage::get()
{
    return 9;
}


uint32 
LSFHandler::LampDetails_IncandescentEquivalent::get()
{
    return 60;
}


uint32 
LSFHandler::LampDetails_MaxLumens::get()
{
    return 620;
}


uint32 
LSFHandler::LampDetails_MinTemperature::get()
{
    return 2700;
}


uint32 
LSFHandler::LampDetails_MaxTemperature::get()
{
    return 9000;
}


uint32 
LSFHandler::LampDetails_ColorRenderingIndex::get()
{
    return 80;
}


Platform::String^ 
LSFHandler::LampDetails_LampID::get()
{
    return m_parentDevice->SerialNumber;
}




// LampState Interface
uint32 
LSFHandler::LampState_Version::get()
{
    return 1;
}


bool 
LSFHandler::LampState_OnOff::get()
{
    // IsOn type is Platform::IBox<bool>^ which means that it can be equal to nullptr, true or false
    // unlike bool which only can be true or false
    if (m_parentDevice->LampState->IsOn != nullptr)
    {
        return m_parentDevice->LampState->IsOn->Value;
    }
    else
    {
        // assume false if nullptr
        return false;
    }
}

void 
LSFHandler::LampState_OnOff::set(bool isOn)
{
    // IsOn type is Platform::IBox<bool>^ which means that it can be equal to nullptr, true or false
    // unlike bool which only can be true or false
    if (m_parentDevice->LampState->IsOn != nullptr &&
        m_parentDevice->LampState->IsOn->Value == isOn)
    {
        return;
    }

    m_parentDevice->LampState->IsOn = isOn;
    notifyLampStateChange();
}


uint32 
LSFHandler::LampState_Brightness::get()
{
    // Brightness type is Platform::IBox<uint32>^ which means that unlike uint32 it can be equal to nullptr
    if (m_parentDevice->LampState->Brightness != nullptr)
    {
        return m_parentDevice->LampState->Brightness->Value;
    }
    else
    {
        // assume 0 if nullptr
        return 0;
    }
}

void
LSFHandler::LampState_Brightness::set(uint32 brightness)
{
    // Brightness type is Platform::IBox<uint32>^ which means that unlike uint32 it can be equal to nullptr
    if (m_parentDevice->LampState->Brightness != nullptr &&
        m_parentDevice->LampState->Brightness->Value == brightness)
    {
        return;
    }

    m_parentDevice->LampState->Brightness = brightness;
    notifyLampStateChange();
}


uint32 
LSFHandler::LampState_Hue::get()
{
    // Hue type is Platform::IBox<uint32>^ which means that unlike uint32 it can be equal to nullptr
    if (m_parentDevice->LampState->Hue != nullptr)
    {
        return m_parentDevice->LampState->Hue->Value;
    }
    else
    {
        // assume 0 if nullptr
        return 0;
    }
}

void 
LSFHandler::LampState_Hue::set(uint32 hue)
{
    // Hue type is Platform::IBox<uint32>^ which means that unlike uint32 it can be equal to nullptr
    if (m_parentDevice->LampState->Hue != nullptr &&
        m_parentDevice->LampState->Hue->Value == hue)
    {
        return;
    }

    m_parentDevice->LampState->Hue = hue;
    notifyLampStateChange();
}


uint32 
LSFHandler::LampState_Saturation::get()
{
    // Saturation type is Platform::IBox<uint32>^ which means that unlike uint32 it can be equal to nullptr
    if (m_parentDevice->LampState->Saturation != nullptr)
    {
        return m_parentDevice->LampState->Saturation->Value;
    }
    else
    {
        // assume 0 if nullptr
        return 0;
    }
}

void
LSFHandler::LampState_Saturation::set(uint32 saturation)
{
    // Saturation type is Platform::IBox<uint32>^ which means that unlike uint32 it can be equal to nullptr
    if (m_parentDevice->LampState->Saturation != nullptr &&
        m_parentDevice->LampState->Saturation->Value == saturation)
    {
        return;
    }

    m_parentDevice->LampState->Saturation = saturation;
    notifyLampStateChange();
}


uint32 
LSFHandler::LampState_ColorTemp::get()
{
    // ColorTemp type is Platform::IBox<uint32>^ which means that unlike uint32 it can be equal to nullptr
    if (m_parentDevice->LampState->ColorTemp != nullptr)
    {
        return m_parentDevice->LampState->ColorTemp->Value;
    }
    else
    {
        // assume 0 if nullptr
        return 0;
    }
}

void
LSFHandler::LampState_ColorTemp::set(uint32 colorTemp)
{
    // ColorTemp type is Platform::IBox<uint32>^ which means that unlike uint32 it can be equal to nullptr
    if (m_parentDevice->LampState->ColorTemp != nullptr &&
        m_parentDevice->LampState->ColorTemp->Value == colorTemp)
    {
        return;
    }

    m_parentDevice->LampState->ColorTemp = colorTemp;
    notifyLampStateChange();
}


uint32 
LSFHandler::TransitionLampState(
    _In_ uint64 Timestamp,
    _In_ BridgeRT::State^ NewState,
    _In_ uint32 TransitionPeriod,
    _Out_ uint32 *LampResponseCode
    )
{
    UNREFERENCED_PARAMETER(Timestamp);
    UNREFERENCED_PARAMETER(TransitionPeriod);

    if (!isSameState(m_parentDevice->LampState, NewState))
    {
        m_parentDevice->LampState = NewState;
        notifyLampStateChange();
    }

    *LampResponseCode = 0;

    return ERROR_SUCCESS;
}


uint32 
LSFHandler::LampState_ApplyPulseEffect(
    _In_ BridgeRT::State^ FromState,
    _In_ BridgeRT::State^ ToState,
    _In_ uint32 Period,
    _In_ uint32 Duration,
    _In_ uint32 NumPulses,
    _In_ uint64 Timestamp,
    _Out_ uint32 *LampResponseCode
    )
{
    UNREFERENCED_PARAMETER(FromState);
    UNREFERENCED_PARAMETER(Period);
    UNREFERENCED_PARAMETER(Duration);
    UNREFERENCED_PARAMETER(NumPulses);
    UNREFERENCED_PARAMETER(Timestamp);

    if (!isSameState(m_parentDevice->LampState, ToState))
    {
        m_parentDevice->LampState = ToState;
        notifyLampStateChange();
    }

    *LampResponseCode = 0;

    return ERROR_SUCCESS;
}


BridgeRT::IAdapterSignal^ 
LSFHandler::LampState_LampStateChanged::get()
{
    return m_parentDevice->GetSignalByName(LAMP_STATE_CHANGED_SIGNAL_NAME);
}


void 
LSFHandler::notifyLampStateChange()
{
    MockAdapter^ adapter = dynamic_cast<MockAdapter^>(m_parentDevice->Parent);
    if (adapter == nullptr)
    {
        return;
    }

    MockAdapterSignal^ signal = m_parentDevice->GetSignalByName(LAMP_STATE_CHANGED_SIGNAL_NAME);
    if (signal == nullptr)
    {
        return;
    }

    adapter->NotifySignalListener(signal);
}


bool 
LSFHandler::isSameState(
    _In_ BridgeRT::State^ currentState,
    _In_ BridgeRT::State^ newState
    )
{
    if (currentState->IsOn != newState->IsOn ||
        currentState->Brightness != newState->Brightness ||
        currentState->Hue != newState->Hue ||
        currentState->Saturation != newState->Saturation ||
        currentState->ColorTemp != newState->ColorTemp
        )
    {
        return false;
    }

    return true;
}